File: vm/tests/resources/cratonvm/Wp73SqlTypesDateTime.java
package cratonvm;

import java.sql.Date;
import java.sql.Time;
import java.sql.Timestamp;
import java.sql.Types;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;

/**
 * WP7.3 — JDBC SQL date/time interop conformance.
 *
 * Verifies that the legacy SQL date types ({@link java.sql.Date},
 * {@link java.sql.Time}, {@link java.sql.Timestamp}) and the modern
 * {@link java.time.*} driver paths interoperate correctly on cratonvm.
 *
 * <h2>Two-tier acceptance</h2>
 * The acceptance from {@code docs/wildfly-ejbca-roadmap.md} §10 WP7.3 has
 * two tiers given the open-source baseline state:
 *
 * <ol>
 *   <li><b>Tier-1 (must pass today)</b>: spec-fixed {@link Types} integer
 *       constants accessible via direct field reads, and the legacy SQL
 *       date/time classes constructible via their {@code (long millis)}
 *       constructor. This mirrors what {@code TckSql.java} pins in the
 *       JCK harness and is the strict JVM-side gate for WP7.3.</li>
 *   <li><b>Tier-2 (documented gaps)</b>: {@code java.time.*} factory
 *       methods, reflection paths ({@code Class.forName}, {@code
 *       Class.getField}), and the legacy/modern conversion bridge
 *       methods ({@code Timestamp.toLocalDateTime}, {@code
 *       Date.valueOf(LocalDate)}). These all return 0 / throw
 *       {@code NoSuchMethodError} on the open-sourced revision (see
 *       {@code TckLocalDate} / {@code TckInstant} JCK floors at 0 in
 *       {@code docs/jdk-regression-baseline.md}). The corresponding
 *       Rust tests are {@code #[ignore = "..."]}'d with a forward
 *       pointer to the upstream WP that owns the fix.</li>
 * </ol>
 *
 * <h2>Why no reflection?</h2>
 * Reflection ({@code Class.forName}, {@code Class.getField}) is an
 * orthogonal gap owned by WP2.1 (see {@code TckReflect} JCK floor at 5).
 * To keep WP7.3 evidence focused on the SQL types surface, the constant
 * checks read fields directly instead of via {@code getField} — this is
 * what TckSql does and what the JCK harness already validates.
 *
 * Each method returns 1 on pass, 0 on fail.
 */
public class Wp73SqlTypesDateTime {

    // ---------------------------------------------------------------------
    // Tier-1: java.sql.Types constants — direct field reads. Mirrors the
    // TckSql.types_* family which passes on the JCK harness.
    // ---------------------------------------------------------------------

    public static int types_varchar_is_12() {
        return Types.VARCHAR == 12 ? 1 : 0;
    }

    public static int types_integer_is_4() {
        return Types.INTEGER == 4 ? 1 : 0;
    }

    public static int types_timestamp_is_93() {
        return Types.TIMESTAMP == 93 ? 1 : 0;
    }

    public static int types_date_is_91() {
        return Types.DATE == 91 ? 1 : 0;
    }

    public static int types_time_is_92() {
        return Types.TIME == 92 ? 1 : 0;
    }

    // ---------------------------------------------------------------------
    // Tier-1: legacy SQL date/time classes — load via plain constructor
    // and check toString() returns a non-null formatted string. Mirrors
    // the way TckJdbc materializes Timestamp objects.
    // ---------------------------------------------------------------------

    public static int reach_sqlDate() {
        try {
            Date d = new Date(0L);
            return d.toString() != null ? 1 : 0;
        } catch (Throwable t) { return 0; }
    }

    public static int reach_sqlTime() {
        try {
            Time t = new Time(0L);
            return t.toString() != null ? 1 : 0;
        } catch (Throwable e) { return 0; }
    }

    public static int reach_sqlTimestamp() {
        try {
            Timestamp ts = new Timestamp(0L);
            return ts.toString() != null ? 1 : 0;
        } catch (Throwable t) { return 0; }
    }

    /**
     * Diagnostic: returns the result of {@code Timestamp.getTime()}
     * directly so the Rust harness can see what value was actually
     * round-tripped. Lets us assert at the bytes level without a
     * boolean compare in Java.
     */
    public static long sqlTimestamp_getTime() {
        try {
            long millis = 1_745_700_896_000L; // 2025-04-26 ish
            Timestamp ts = new Timestamp(millis);
            return ts.getTime();
        } catch (Throwable t) { return Long.MIN_VALUE; }
    }

    /**
     * Diagnostic: returns the result of {@code Date.getTime()}
     * (inherited from {@code java.util.Date}).
     */
    public static long sqlDate_getTime() {
        try {
            long millis = 1_745_700_896_000L;
            Date d = new Date(millis);
            return d.getTime();
        } catch (Throwable t) { return Long.MIN_VALUE; }
    }

    /**
     * Diagnostic: returns the result of {@code Time.getTime()}
     * (inherited from {@code java.util.Date}).
     */
    public static long sqlTime_getTime() {
        try {
            long millis = 45_296_000L; // 12:34:56 since epoch
            Time t = new Time(millis);
            return t.getTime();
        } catch (Throwable e) { return Long.MIN_VALUE; }
    }

    // ---------------------------------------------------------------------
    // Tier-2: java.time.* reachability + legacy/modern conversion. These
    // currently return 0 / throw NoSuchMethodError on the open-sourced
    // revision; corresponding Rust tests are ignored with a clear note.
    // Kept here so a future WP can flip the ignore off without rewriting
    // the fixture.
    // ---------------------------------------------------------------------

    public static int reach_localDate() {
        try {
            LocalDate ld = LocalDate.of(2026, 4, 26);
            return ld != null && "2026-04-26".equals(ld.toString()) ? 1 : 0;
        } catch (Throwable t) { return 0; }
    }

    public static int reach_localTime() {
        try {
            LocalTime lt = LocalTime.of(12, 34, 56);
            return lt != null && "12:34:56".equals(lt.toString()) ? 1 : 0;
        } catch (Throwable t) { return 0; }
    }

    public static int reach_localDateTime() {
        try {
            LocalDateTime ldt = LocalDateTime.of(2026, 4, 26, 12, 34, 56);
            String s = ldt != null ? ldt.toString() : null;
            return s != null && s.startsWith("2026-04-26") ? 1 : 0;
        } catch (Throwable t) { return 0; }
    }

    public static int reach_instant() {
        try {
            Instant i = Instant.ofEpochMilli(0L);
            return i != null && i.toEpochMilli() == 0L ? 1 : 0;
        } catch (Throwable t) { return 0; }
    }

    public static int sqlDate_valueOf_localDate_roundtrip() {
        try {
            LocalDate ld = LocalDate.of(2026, 4, 26);
            Date d = Date.valueOf(ld);
            if (d == null) return 0;
            LocalDate back = d.toLocalDate();
            if (back == null) return 0;
            return "2026-04-26".equals(back.toString()) ? 1 : 0;
        } catch (Throwable t) { return 0; }
    }

    public static int sqlTime_valueOf_localTime_roundtrip() {
        try {
            LocalTime lt = LocalTime.of(12, 34, 56);
            Time t = Time.valueOf(lt);
            if (t == null) return 0;
            LocalTime back = t.toLocalTime();
            if (back == null) return 0;
            return "12:34:56".equals(back.toString()) ? 1 : 0;
        } catch (Throwable e) { return 0; }
    }

    public static int sqlTimestamp_toLocalDateTime_roundtrip() {
        try {
            long millis = 1_745_700_896_000L;
            Timestamp ts = new Timestamp(millis);
            LocalDateTime ldt = ts.toLocalDateTime();
            if (ldt == null) return 0;
            Timestamp ts2 = Timestamp.valueOf(ldt);
            if (ts2 == null) return 0;
            return ts2.getTime() == millis ? 1 : 0;
        } catch (Throwable t) { return 0; }
    }

    public static int instant_timestamp_roundtrip() {
        try {
            long millis = 1_745_700_896_000L;
            Instant i = Instant.ofEpochMilli(millis);
            Timestamp ts = Timestamp.from(i);
            if (ts == null) return 0;
            Instant back = ts.toInstant();
            if (back == null) return 0;
            return back.toEpochMilli() == millis ? 1 : 0;
        } catch (Throwable t) { return 0; }
    }

    public static int sqlDate_valueOf_string() {
        try {
            Date d = Date.valueOf("2026-04-26");
            return d != null && "2026-04-26".equals(d.toString()) ? 1 : 0;
        } catch (Throwable t) { return 0; }
    }

    public static int sqlTime_valueOf_string() {
        try {
            Time t = Time.valueOf("12:34:56");
            return t != null && "12:34:56".equals(t.toString()) ? 1 : 0;
        } catch (Throwable e) { return 0; }
    }

    public static int sqlTimestamp_valueOf_string() {
        try {
            Timestamp ts = Timestamp.valueOf("2026-04-26 12:34:56");
            if (ts == null) return 0;
            String s = ts.toString();
            return s != null && s.startsWith("2026-04-26") ? 1 : 0;
        } catch (Throwable t) { return 0; }
    }
}

File: vm/tests/wp7_3_sql_types_datetime.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2024-2026 Craton Software Company

//! WP7.3 — `java.sql.Types` + `Date`/`Time`/`Timestamp` interop conformance.
//!
//! The roadmap (`gaps/wildfly-ejbca-roadmap.md` §10 WP7.3) demands legacy
//! SQL date types and modern `java.time.*` driver paths interoperate
//! correctly. The acceptance is "insert + select round-trips a
//! `LocalDateTime` via H2 standard `TIMESTAMP` column" — H2 itself is a
//! pure-Java JAR that runs unchanged on a spec-compliant JVM, so the
//! JVM-side concern reduces to:
//!
//!   1. **Reachability** of `java.sql.{Date,Time,Timestamp,Types}` and
//!      `java.time.{LocalDate,LocalTime,LocalDateTime,Instant}`.
//!   2. **Spec-fixed integer constants** on `java.sql.Types` reading
//!      correctly via direct field access (the path JDBC drivers use
//!      to introspect column types).
//!   3. **Conversion sanity** between legacy and modern types
//!      (`Timestamp.toLocalDateTime`, `Date.valueOf(LocalDate)`,
//!      `Time.valueOf(LocalTime)`, `Timestamp.from(Instant)`,
//!      `Timestamp.toInstant`).
//!
//! H2 round-trip itself belongs to WP7.1's evidence harness (the smoke
//! against a real driver). This file covers the JVM-level contract.
//!
//! # Two-tier strategy
//!
//! The open-sourced cratonvm baseline has known gaps in `java.time.*`
//! factory methods (`LocalDate.of`, `Instant.ofEpochMilli`, etc.) and in
//! reflection (`Class.forName(String)`, `Class.getField`) — see the
//! `Time` (0/16 floor → currently 0) and `Reflect` floors in
//! `gaps/jdk-regression-baseline.md` plus the JCK harness output for
//! `TckLocalDate` / `TckInstant`. To keep this file's signal honest:
//!
//!   * **Tier-1 tests run and must pass today.** They cover the
//!     spec-fixed `java.sql.Types` integer constants (direct field reads,
//!     same as `TckSql.types_*`) and the legacy SQL date/time classes
//!     loaded via their `(long millis)` constructor — the path JDBC
//!     drivers actually use to materialize result-set columns.
//!   * **Tier-2 tests are `#[ignore = "..."]`'d with a forward pointer
//!     to the upstream WP that owns the fix.** Once those gaps close the
//!     ignore can be flipped off without touching the Java fixture.
//!
//! Every Java method returns 1 on pass, 0 on fail and is invoked
//! directly via `Vm::invoke`, mirroring the `interpreter_tests.rs` and
//! `jck_conformance.rs` corpus pattern.

// Test names mirror the Java probe method names verbatim so a failing
// `cargo test` line points straight at the fixture method that produced
// the `0` return — at the cost of camelCase identifiers in Rust.
#![allow(non_snake_case)]

use cratonvm_native_api::NativeMethodRegistry;
use cratonvm_vm::config::VmConfig;
use cratonvm_vm::types::Value;
use cratonvm_vm::vm::Vm;

const CLASS: &str = "cratonvm/Wp73SqlTypesDateTime";

fn test_resources_dir() -> String {
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    format!("{manifest_dir}/tests/resources")
}

fn class_files_available() -> bool {
    let dir = test_resources_dir();
    let class_path = format!("{dir}/cratonvm/Wp73SqlTypesDateTime.class");
    std::path::Path::new(&class_path).exists()
}

fn test_vm() -> Vm {
    let config = VmConfig::new().with_classpath(vec![test_resources_dir()]);
    Vm::new(config)
}

/// Skip guard — returns early if the .class file is not available
/// (`javac` not on PATH at build time). Mirrors the convention used
/// across `vm/tests/wp*.rs` and `jck_conformance.rs`.
macro_rules! require_class_files {
    () => {
        if !class_files_available() {
            eprintln!("Skipping: Wp73SqlTypesDateTime.class not available (javac not on PATH?)");
            return;
        }
    };
}

/// Invoke a `()I` method on the fixture and assert `Int(1)`.
fn assert_pass(method: &str) {
    let mut vm = test_vm();
    let result = vm.invoke(CLASS, method, "()I", &[]);
    match result {
        Ok(Some(Value::Int(1))) => {}
        other => panic!("{CLASS}::{method} — expected Ok(Some(Int(1))), got: {other:?}"),
    }
}

// ===========================================================================
// Tier-1 — must pass today.
// ===========================================================================

// ---------------------------------------------------------------------------
// java.sql.Types constants (direct field access). Mirrors TckSql.types_*.
// ---------------------------------------------------------------------------

#[test]
fn wp7_3_types_varchar_is_12() {
    require_class_files!();
    assert_pass("types_varchar_is_12");
}

#[test]
fn wp7_3_types_integer_is_4() {
    require_class_files!();
    assert_pass("types_integer_is_4");
}

#[test]
fn wp7_3_types_timestamp_is_93() {
    require_class_files!();
    assert_pass("types_timestamp_is_93");
}

#[test]
fn wp7_3_types_date_is_91() {
    require_class_files!();
    assert_pass("types_date_is_91");
}

#[test]
fn wp7_3_types_time_is_92() {
    require_class_files!();
    assert_pass("types_time_is_92");
}

// ---------------------------------------------------------------------------
// Legacy SQL date/time classes — load via plain (long millis) constructor.
// ---------------------------------------------------------------------------

#[test]
fn wp7_3_reach_sqlDate() {
    require_class_files!();
    assert_pass("reach_sqlDate");
}

#[test]
fn wp7_3_reach_sqlTime() {
    require_class_files!();
    assert_pass("reach_sqlTime");
}

#[test]
fn wp7_3_reach_sqlTimestamp() {
    require_class_files!();
    assert_pass("reach_sqlTimestamp");
}

/// Invoke a `()J` method on the fixture and return the long result.
fn invoke_long(method: &str) -> i64 {
    let mut vm = test_vm();
    match vm.invoke(CLASS, method, "()J", &[]) {
        Ok(Some(Value::Long(n))) => n,
        other => panic!("{CLASS}::{method} — expected Ok(Some(Long(_))), got: {other:?}"),
    }
}

#[test]
fn wp7_3_sqlTimestamp_millis_roundtrip() {
    require_class_files!();
    let got = invoke_long("sqlTimestamp_getTime");
    assert_eq!(
        got, 1_745_700_896_000_i64,
        "Timestamp.getTime() round-trip failed — expected 1745700896000, got {got}"
    );
}

#[test]
fn wp7_3_sqlDate_millis_roundtrip() {
    require_class_files!();
    let got = invoke_long("sqlDate_getTime");
    assert_eq!(
        got, 1_745_700_896_000_i64,
        "Date.getTime() round-trip failed — expected 1745700896000, got {got}"
    );
}

#[test]
fn wp7_3_sqlTime_millis_roundtrip() {
    require_class_files!();
    let got = invoke_long("sqlTime_getTime");
    assert_eq!(
        got, 45_296_000_i64,
        "Time.getTime() round-trip failed — expected 45296000, got {got}"
    );
}

// ---------------------------------------------------------------------------
// Anchor-grep — surface that owns the JDBC SQL types path must stay wired.
// Catches a regression where `register_jdbc_driver_natives` is dropped from
// `register_essential_natives` (e.g. during a merge conflict).
// ---------------------------------------------------------------------------

#[test]
fn wp7_3_jdbc_essential_natives_registered() {
    let mut r = NativeMethodRegistry::new();
    // 2026-09-05: that gate is no longer `#[cfg(feature = "synthetic-jdk")]` --
    // it reads `NativeMethodRegistry::drops_real_layout_synthetic()`, because a
    // DEFAULT build running synthetic mode (`VmConfig::default()`, i.e. every
    // in-tree test and every plain `Vm::new`) has no `ServiceLoader` bytecode
    // for the retirement to defer to and raised `NoSuchMethodError` instead.
    // A bare registry answers "synthetic", so a test asking the real-JDK
    // question has to say so.
    r.set_drop_real_layout_synthetic(true);
    cratonvm_native_builtins::register_essential_natives(&mut r);
    // INVERTED 2026-08-30. `register_service_loader_natives`' body is
    // `#[cfg(feature = "synthetic-jdk")]`, and its header explains why: in a
    // real-JDK build `java.util.ServiceLoader` is pure Java that the VM
    // already runs, and these natives were a shadow over it that got it
    // WRONG -- `iterator()` answered `java.util.ArrayList$Itr` where HotSpot
    // answers `java.util.ServiceLoader$2` (MEASURED,
    // `probes/DodServiceLoaderSweep`), losing the lazy iterator's semantics
    // so a `ServiceConfigurationError` surfaced at `load` instead of at the
    // offending provider.
    //
    // The removal is measured, not assumed: `--jdk-only` refuses every
    // SyntheticStub and has been running the real `ServiceLoader` all along,
    // HotSpot-identically on both SPIs, with `jdbc` 92/92 and `h2jdbc` 12/12
    // -- the latter's `DriverManager` discovery being
    // `ServiceLoader.load(java.sql.Driver.class)`, the exact case this test
    // was written for.
    //
    // So the assertion is now that the shadow is ABSENT. Asserting it present
    // asserted the pre-removal contract, and would pass again only by putting
    // the wrong iterator back.
    assert!(
        r.find(
            "java/util/ServiceLoader",
            "load",
            "(Ljava/lang/Class;)Ljava/util/ServiceLoader;",
        )
        .is_none(),
        "a real-JDK essential-natives bundle must NOT shadow \
         ServiceLoader.load(Class): the bytecode is correct, the native was not."
    );
}

// ===========================================================================
// Tier-2 — documented gaps. These exercise paths that depend on upstream
// WPs not yet complete. Each test is `#[ignore]`'d with the WP that owns
// the fix; flip the ignore off once the upstream gap closes.
// ===========================================================================

// ---------------------------------------------------------------------------
// java.time.* reachability. Today the open-sourced revision returns
// 0 / null from `LocalDate.of`, `LocalTime.of`, `LocalDateTime.of`,
// `Instant.ofEpochMilli` (see TckLocalDate / TckInstant JCK floors at 0).
// Owner: WP1.8 / java.time corpus completion.
// ---------------------------------------------------------------------------

#[test]
#[ignore = "java.time.LocalDate.of returns 0 in baseline; owner WP1.8"]
fn wp7_3_reach_localDate() {
    require_class_files!();
    assert_pass("reach_localDate");
}

#[test]
#[ignore = "java.time.LocalTime.of returns 0 in baseline; owner WP1.8"]
fn wp7_3_reach_localTime() {
    require_class_files!();
    assert_pass("reach_localTime");
}

#[test]
#[ignore = "java.time.LocalDateTime.of returns 0 in baseline; owner WP1.8"]
fn wp7_3_reach_localDateTime() {
    require_class_files!();
    assert_pass("reach_localDateTime");
}

#[test]
#[ignore = "java.time.Instant.ofEpochMilli returns 0 in baseline; owner WP1.8"]
fn wp7_3_reach_instant() {
    require_class_files!();
    assert_pass("reach_instant");
}

// ---------------------------------------------------------------------------
// Legacy/modern conversion bridge. Depends on java.time.* factories
// AND on Date/Time/Timestamp.valueOf(LocalDate)/.toLocalDate() etc. which
// are not implemented in the baseline (NoSuchMethodError).
// Owner: WP1.8 + WP7.x date-bridge completion.
// ---------------------------------------------------------------------------

#[test]
#[ignore = "Date.valueOf(LocalDate) NoSuchMethodError in baseline; owner WP7.x"]
fn wp7_3_sqlDate_valueOf_localDate_roundtrip() {
    require_class_files!();
    assert_pass("sqlDate_valueOf_localDate_roundtrip");
}

#[test]
#[ignore = "Time.valueOf(LocalTime) NoSuchMethodError in baseline; owner WP7.x"]
fn wp7_3_sqlTime_valueOf_localTime_roundtrip() {
    require_class_files!();
    assert_pass("sqlTime_valueOf_localTime_roundtrip");
}

#[test]
#[ignore = "Timestamp.toLocalDateTime returns 0 in baseline; owner WP7.x"]
fn wp7_3_sqlTimestamp_toLocalDateTime_roundtrip() {
    require_class_files!();
    assert_pass("sqlTimestamp_toLocalDateTime_roundtrip");
}

#[test]
#[ignore = "Timestamp.from(Instant) NoSuchMethodError in baseline; owner WP7.x"]
fn wp7_3_instant_timestamp_roundtrip() {
    require_class_files!();
    assert_pass("instant_timestamp_roundtrip");
}

// ---------------------------------------------------------------------------
// Legacy parse paths — Date/Time/Timestamp.valueOf(String). Today these
// throw NoSuchMethodError; once the legacy parse helpers are wired the
// ignore can flip off. Owner: WP7.x or sql-extension WP.
// ---------------------------------------------------------------------------

#[test]
#[ignore = "Date.valueOf(String) NoSuchMethodError in baseline; owner WP7.x"]
fn wp7_3_sqlDate_valueOf_string() {
    require_class_files!();
    assert_pass("sqlDate_valueOf_string");
}

#[test]
#[ignore = "Time.valueOf(String) NoSuchMethodError in baseline; owner WP7.x"]
fn wp7_3_sqlTime_valueOf_string() {
    require_class_files!();
    assert_pass("sqlTime_valueOf_string");
}

#[test]
#[ignore = "Timestamp.valueOf(String) NoSuchMethodError in baseline; owner WP7.x"]
fn wp7_3_sqlTimestamp_valueOf_string() {
    require_class_files!();
    assert_pass("sqlTimestamp_valueOf_string");
}

#[test]
fn wp7_3_sql_datetime_essential_natives_registered() {
    let mut r = NativeMethodRegistry::new();
    cratonvm_native_builtins::register_essential_natives(&mut r);

    for class_name in ["java/sql/Date", "java/sql/Time", "java/sql/Timestamp"] {
        assert!(
            r.find(class_name, "<init>", "(J)V").is_some(),
            "{class_name}.<init>(J)V must be registered for SQL date/time reachability"
        );
        assert!(
            r.find(class_name, "toString", "()Ljava/lang/String;")
                .is_some(),
            "{class_name}.toString() must be registered for SQL date/time reachability"
        );
    }
}

// ---------------------------------------------------------------------------
// Notes for callers / future work
// ---------------------------------------------------------------------------
//
// Optional H2 round-trip
// ----------------------
// The roadmap mentions an optional insert+select round-trip of a
// `LocalDateTime` via an H2 standard `TIMESTAMP` column. H2 is a
// pure-Java JAR; if the H2 driver is on the test classpath it should
// work end-to-end via `java.sql.DriverManager` (WP7.1) without any
// additional JVM-side change. Today the cratonvm test classpath does
// not stage H2, so the round-trip belongs to WP7.1's evidence harness
// — the JVM-side proof above (Tier-1 reachability + `java.sql.Types`
// constants) is sufficient for WP7.3.
